src/lib/transition_frontier/full_frontier/full_frontier.ml

```diff
@@ -0,0 +1,566 @@
+          num_breadcrumbs_removed ;
+        Gauge.set Transition_frontier.recently_finalized_staged_txns
+          num_finalized_staged_txns ;
+        Counter.inc Transition_frontier.finalized_staged_txns
+          num_finalized_staged_txns ;
+        Gauge.set Transition_frontier.root_snarked_ledger_accounts
+          num_root_snarked_ledger_accounts ;
+        Gauge.set Transition_frontier.root_snarked_ledger_total_currency
+          root_snarked_ledger_total_currency ;
+        Counter.inc_one Transition_frontier.root_transitions ;
+        Gauge.set Transition_frontier.slot_fill_rate
+          (blockchain_length /. global_slot))
+  | _ ->
+      ()
+
+let apply_diffs t diffs =
+  let open Root_identifier.Stable.Latest in
+  Logger.trace t.logger ~module_:__MODULE__ ~location:__LOC__
+    "Applying %d diffs to full frontier (%s --> ?)" (List.length diffs)
+    (Frontier_hash.to_string t.hash) ;
+  let local_state_was_synced_at_start =
+    Consensus.Hooks.required_local_state_sync
+      ~consensus_state:(Breadcrumb.consensus_state (best_tip t))
+      ~local_state:t.consensus_local_state
+    |> Option.is_none
+  in
+  let new_root =
+    List.fold diffs ~init:None ~f:(fun prev_root (Diff.Full.E.E diff) ->
+        let mutant, new_root = apply_diff t diff in
+        t.hash <- Frontier_hash.merge_diff t.hash (Diff.to_lite diff) mutant ;
+        update_metrics_with_diff t diff ;
+        match new_root with
+        | None ->
+            prev_root
+        | Some state_hash ->
+            Some {state_hash; frontier_hash= t.hash} )
+  in
+  Logger.trace t.logger ~module_:__MODULE__ ~location:__LOC__
+    "Reached state %s after applying diffs to full frontier"
+    (Frontier_hash.to_string t.hash) ;
+  Debug_assert.debug_assert (fun () ->
+      match
+        Consensus.Hooks.required_local_state_sync
+          ~consensus_state:
+            (Breadcrumb.consensus_state
+               (Hashtbl.find_exn t.table t.best_tip).breadcrumb)
+          ~local_state:t.consensus_local_state
+      with
+      | Some jobs ->
+          (* But if there wasn't sync work to do when we started, then there shouldn't be now. *)
+          if local_state_was_synced_at_start then (
+            Logger.fatal t.logger
+              "after lock transition, the best tip consensus state is out of \
+               sync with the local state -- bug in either \
+               required_local_state_sync or frontier_root_transition."
+              ~module_:__MODULE__ ~location:__LOC__
+              ~metadata:
+                [ ( "sync_jobs"
+                  , `List
+                      ( Non_empty_list.to_list jobs
+                      |> List.map ~f:Consensus.Hooks.local_state_sync_to_yojson
+                      ) )
+                ; ( "local_state"
+                  , Consensus.Data.Local_state.to_yojson
+                      t.consensus_local_state )
+                ; ("tf_viz", `String (visualize_to_string t)) ] ;
+            assert false )
```

Comment:
can you failwith with a message instead of assert-false? Will save the crash reader at least 30sec if this ever happens.